app/helpers/browse_tag_changes_helper.rb

```diff
@@ -0,0 +1,134 @@
+    }.fetch(change_type, "")
+  end
+
+  def format_tag_value_with_change(key, change_info)
+    case change_info[:type]
+    when :added
+      tag.ins(format_value(key, change_info[:current], :skip_wikidata_preview => true))
+    when :unmodified
+      # For added and unmodified show the current value without the wikidata preview
+      format_value(key, change_info[:current], :skip_wikidata_preview => true)
+    when :modified
+      # Return array of two values for the two rows that will be created
+      [
+        tag.del(format_value(key, change_info[:previous], :skip_wikidata_preview => true)),
+        tag.ins(format_value(key, change_info[:current], :skip_wikidata_preview => true))
+      ]
+    when :removed
+      tag.del(format_value(key, change_info[:previous] || "", :skip_wikidata_preview => true))
+    when nil
+      # For unknown versioning show the current value with the wikidata preview
+      format_value(key, change_info[:current] || "", :skip_wikidata_preview => false)
+    end
+  end
+
+  def get_change_indicator_text(change_type)
+    case change_type
+    when :added
+      "+"
+    when :removed
+      "−"
+    else
+      " "
+    end
+  end
+
+  def get_indicator_cell_class(change_type)
+    case change_type
+    when :added
+      "diff-indicator-cell diff-added"
+    when :removed
+      "diff-indicator-cell diff-removed"
+    when :unmodified
+      "diff-indicator-cell diff-unmodified"
+    else
+      "diff-indicator-cell"
+    end
+  end
+
+  def format_tag_row_with_change(key, change_info)
+    case change_info[:type]
+    when :modified
+      # Generate two rows for modified tags, using tag-removed and tag-added classes on rows
+      value_cells = format_tag_value_with_change(key, change_info)
+      [
+        tag.tr(:class => "tag-removed") do
+          safe_join([
+                      tag.th(format_key(key), :class => "py-1 border-secondary-subtle table-secondary fw-normal diff-key-modified", :rowspan => 2),
+                      tag.td(get_change_indicator_text(:removed), :class => get_indicator_cell_class(:removed)),
+                      tag.td(value_cells[0], :class => "py-1 diff-cell")
+                    ])
+        end,
+        tag.tr(:class => "tag-added") do
+          safe_join([
+                      tag.td(get_change_indicator_text(:added), :class => get_indicator_cell_class(:added)),
+                      tag.td(value_cells[1], :class => "py-1 diff-cell")
+                    ])
+        end
+      ]
+    else
+      # Generate single row for other change types
+      cells = [tag.th(format_key(key), :class => "py-1 border-secondary-subtle table-secondary fw-normal")]
+
+      # Only add diff indicator cell if we're in a history/diff context
+      cells << tag.td(get_change_indicator_text(change_info[:type]), :class => get_indicator_cell_class(change_info[:type])) if change_info[:type]
+
+      cells << tag.td(format_tag_value_with_change(key, change_info), :class => "py-1")
```

Comment:
Why break this down into multiple statements that append to a temporary array when the modified version just does it all inline?